vm/ubpf_vm.c

```diff
@@ -595,20 +599,61 @@ int
 ubpf_run(struct ubpf_vm *vm, void *mem, size_t mem_len, bool jit,
     uint64_t *bpf_return_value, char **errmsg)
 {
+    int cntrfd = 0;
+    ubpf_jit_fn fn;
+
+    if (vm->inst_cnt) {
+        cntrfd = setup_instruction_counter();
+        if (cntrfd == -1) {
+            vm->error_printf(stderr, "uBPF warning: unable to setup instruction counter\n");
+        }
+    }
+
     if (jit) {
-        ubpf_jit_fn fn = ubpf_compile(vm, errmsg);
+        if (vm->inst_cnt) {
+            enable_instruction_count(cntrfd);
+        }
+        fn = ubpf_compile(vm, errmsg);
+        if (vm->inst_cnt) {
```

Comment:
This should be `if (vm->inst_cnt && cntrfd != -1)`, otherwise we will end up reaching the ubpf_error()